docker/docker_client.go

```diff
@@ -208,8 +209,19 @@ func newDockerClientWithDetails(sys *types.SystemContext, registry, username, pa
 		return nil, err
 	}
 
+	// If specified in the SystemContext, use its InsecureSkipVerify flag.
+	// Otherwise, query if the remoteName has a correspoding registry in
+	// the registry config and set the flag accordingly.
 	if sys != nil && sys.DockerInsecureSkipTLSVerify {
 		tr.TLSClientConfig.InsecureSkipVerify = true
+	} else if sys != nil {
```

Comment:
maybe I need more tea, but if the user had set DockerInsecureSkipTLSVerify to false, meaning to use tls-verify, should we be potentially setting InsecureSkipVerify to true if that's the setting in reg.Insecure?

Reply:
The idea behind this code is to allow users of the API to opt into full verification even if the `registries.conf` sets the given registry to be insecure (see discussion at https://github.com/containers/image/pull/464#issuecomment-397744605).

@mtrmac proposed this semantic and he usually has a good intuition when it comes to such details. The nasty thing with booleans is that there is no way to check if a given item is "unset". Some tristate would be much nicer to resolve such issues.

Reply:
Right if I as a user of podman specify --tlsverify=TRUE, that means ignore the settings in the registries.conf file and only use https.

Reply:
> The idea behind this code is to allow users of the API to opt into full verification even if the `registries.conf` sets the given registry to be insecure

I can’t see how this code is doing that, unless you mean “by setting `sys == nil`”; but `types.SystemContext` documentation says that `nil` and `&types.SystemContext{}` must be treated equivalently everywhere, so that conditional would not be appropriate.

If we need a tri-state, then let’s actually set one up.  Probably an `enum` undefined/insecure/secure instead of a `*bool`—and enum will require changes in all users, but it’s easier to set an enum than a `*bool` in a struct initializer.

Or maybe add a `bool DockerForceSecureTLSVerification`, and fail if both this and `DockerInsecureSkipTLSVerify` is set — that would be compatible, but it is _very_ ugly.

But…

---

… I’m not actually convinced we _need_ this “force verification on even if the config turns it off” override; I brought it up primarily because the `--tls-verify` behavior of libpod/buildah works like that.  In which situations would the flag be used like that, though?

The only use I can think of is setting up certificates for a previously insecure registry, to check whether it is safe to change the configuration; in that case it would be better to create the new configuration, and test with `SystemContext.SystemRegistriesConfPath` pointing to that new configuration file; that tests _both_ the TLS setup of the registry and correctness of the contemplated file edits.

(It is specifically not very reasonable for the machine admin to set a registry to insecure, and one user using `--tls-verify=true` because the user disagrees, for whatever reason: it’s difficult to imagine what reason that could be, and it’s way too easy to forget to use `--tls-verify=true`, or to run a third-party script which does not use it.)

---

Yet again, if the flag syntax is `--tls-verify=false`, then `--tls-verify=true` needs to do _something_, and doing no verification (when the config file says so) would seem pretty surprising.

Reply:
> I can’t see how this code is doing that, unless you mean “by setting sys == nil”; but types.SystemContext documentation says that nil and &types.SystemContext{} must be treated equivalently everywhere, so that conditional would not be appropriate.

Good catch, thanks!

> Yet again, if the flag syntax is --tls-verify=false, then --tls-verify=true needs to do something, and doing no verification (when the config file says so) would seem pretty surprising.

Given Skopeo has a similar flag for source and destination registries, I feel that we should follow the tri-state idea. Then we can also error out if the `registries.conf` and SystemContext{} are in conflict.

@mtrmac If you agree, I will update the PR accordignly. I'd make the tri-state generic with /undefined/set/unset/ to make it applicable for other use-cases.

Reply:
> Given Skopeo has a similar flag for source and destination registries, I feel that we should follow the tri-state idea.

ACK.

> Then we can also error out if the registries.conf and SystemContext{} are in conflict.

So far CLI flags, and therefore `SystemContext`, have been overriding the configuration files; I’d prefer to keep that model instead of failing on inconsistency.

I can understand the argument for forbidding `--tls-verify=false` if the config file says to use TLS, but we have had the option without the semantics, so changing it is a bit unattractive. Also, we still allow callers to change the config file path, so enforcing consistency would not give the system admin any reliable enforcement mechanism.

As for forbidding `--tls-verify=true` if the config file says not to use TLS, that would entirely negate the point of introducing the tri-state mechanism: `true` is a default for registries not in the config file, and for registries in the config file, the `--tls-verify` value would have to match, so `--tls-verify=true` could never be used in a way that would make any difference.

Reply:
> So far CLI flags, and therefore SystemContext, have been overriding the configuration files; I’d prefer to keep that model instead of failing on inconsistency.

That makes things much nicer and will delay a heart attack by many years.

### toml and custom types
I have been looking into the tri-state idea and it turns out that toml makes it a bit hard/impossible to implement a tri-state type the way I imagined it (and @mtrmac I think as well). By all means, I wanted to allow users to keep using `true` and `false` in the configuration and to map it to a `type TriState int` when unmarshalling.

It turns out that's not possible with toml (and the available toml go libs). Both, `true` and `false` are always mapped to a `bool` regardless if we specified an unmarshal method for `TriState` and hence always leads to a type-conversion error (bool <-> int). I've been trying hard and skimmed old issues etc. and came to the conclusion that there is no user-friendly way to make a simple TriState type that can be used in toml, CLI flags, etc. Note, that this is possible when using `encoding.json` though. There's more such peculiarities when using toml (int <-> float when "." is missing etc.).

That being sad...
> So far CLI flags, and therefore SystemContext, have been overriding the configuration files; I’d prefer to keep that model instead of failing on inconsistency.

... makes the upper issue much simpler as we only have to care about `SystemContext.DockerInsecureSkipTLSVerify` and how we can check if the default or a user-specified value is used.

@mtrmac Hence, I suggest to make only `DockerInsecureSkipTLSVerify` a `TriState` or `TriBool` and leave the configuration as is. WDYT? (I could have updated the commit with those changes, but I figured that the upper issue is worth discussing before.)

Reply:
> @mtrmac Hence, I suggest to make only DockerInsecureSkipTLSVerify a TriState or TriBool and leave the configuration as is. 

I agree.  (`OptionalBool` maybe?)

We genuinely have three kinds of semantics for the CLI options: “force `true`” / “force `false`” / “use configuration files”.

We don’t _really_ have three kinds of semantics for the configuration: we have “configured `true`” / “configured `false`” / “use default”, but immediately after parsing the configuration, “use default” amounts to the same thing as “configured `false`”; we don’t need to represent “use default” anywhere outside parsing a single `Registry` / `Mirror` entry (and even in there, probably only implicitly).

Reply:
> I agree. (OptionalBool maybe?)
+1